File: ExtremeParser/Global/FunctionParameters.vb
Public Class FunctionParameters

    Public ReadOnly Property ParamsText
        Get
            Return _paramText
        End Get
    End Property

    Public Integers As New Dictionary(Of String, String)
    Public Arrays As New Dictionary(Of String, String)
    Public Floats As New Dictionary(Of String, String)
    Public PawnDoc As PawnDocParser

    Private _othersVar As New Dictionary(Of String, String) 'A function is created to handle this.
    Private _paramText As String

    Public Enum VarTypes
        TypeFloat
        TypeInteger
        TypeArray
        TypeOther
        TypeTagged
    End Enum
    Public Shared Function GetVarType(var As String) As varTypes
        If var.StartsWith("Float:") Then
            Return varTypes.TypeFloat
        ElseIf var.EndsWith("]") And var.Contains("[") Then
            Return varTypes.TypeArray
        ElseIf var.Contains(":") Then
            Return varTypes.TypeTagged
        Else
            Return varTypes.TypeInteger
        End If
    End Function

    Public Sub New(params As String)
        Dim trimmedParams = params.Replace(" ", "")
        _paramText = trimmedParams
        Dim prms As String() = trimmedParams.Split(",")
        For Each str As String In prms
            If str = "" Or str = Nothing Then Continue For

            Dim def As String = ""
            If str.Contains("=") Then def = str.Substring(str.IndexOf("=")).Trim()

            If GetVarType(str) = VarTypes.TypeFloat Then
                str = str.Remove(0, 6)
                Floats.Add(str, def)
            ElseIf GetVarType(str) = VarTypes.TypeArray Then

                Dim done As Boolean = False
                While done = False
                    If str.Contains("[") And str.Contains("]") Then
                        str = str.Remove(str.IndexOf("["), (str.IndexOf("]") - str.IndexOf("[")) + 1)
                    Else
                        done = True
                    End If
                End While

                Arrays.Add(str, def)
            ElseIf GetVarType(str) = VarTypes.TypeTagged Then
                _othersVar.Add(str, def)
            Else 'Its an integer obv..
                Integers.Add(str, def)
            End If
        Next
    End Sub

    Public Function Others(tag As String) As List(Of String)
        Return (From str In _othersVar.Keys Select vars = _othersVar(str).Split(":", 2, StringSplitOptions.None) Where vars(0) = tag Select vars(1)).ToList()
    End Function

    Public Enum ReturnType
        AsEnum
        AsString
    End Enum
    Public Function GetParameterType(paramName As String, returnType As returnType)
        Dim ret As varTypes = getVarType(paramName)

        If returnType = returnType.AsEnum Then
            Return ret
        Else
            If ret = varTypes.TypeArray Then
                Return "Array"
            ElseIf ret = varTypes.TypeFloat
                Return "Float"
            ElseIf ret = varTypes.TypeInteger
                Return "Integer"
            ElseIf ret = varTypes.TypeOther
                Return "Other"
            ElseIf ret = varTypes.TypeTagged
                Return "Tagged"
            End If
        End If
        Return Nothing
    End Function
End Class

File: ExtremeParser/Global/FunctionParameters.cs
using System;
using System.Collections.Generic;
using System.Linq;
using ExtremeParser.Parsers;

namespace ExtremeParser.Global
{
    public class FunctionParameters
    {

        public dynamic ParamsText => _paramText;

        public Dictionary<string, string> Integers = new Dictionary<string, string>();
        public Dictionary<string, string> Arrays = new Dictionary<string, string>();
        public Dictionary<string, string> Floats = new Dictionary<string, string>();
        public PawnDocParser PawnDoc;

        private Dictionary<string, string>
            _othersVar = new Dictionary<string, string>(); //A function is created to handle this.

        private string _paramText;

        public enum VarTypes
        {
            TypeFloat,
            TypeInteger,
            TypeArray,
            TypeOther,
            TypeTagged
        }

        public static VarTypes GetVarType(string var)
        {
            if (var.StartsWith("Float:"))
            {
                return VarTypes.TypeFloat;
            }

            if (var.EndsWith("]") && var.Contains("["))
            {
                return VarTypes.TypeArray;
            }

            if (var.Contains(":"))
            {
                return VarTypes.TypeTagged;
            }

            return VarTypes.TypeInteger;
        }

        public FunctionParameters(string @params)
        {
            var trimmedParams = @params.Replace(" ", "");
            _paramText = trimmedParams;
            string[] prms = trimmedParams.Split(",".ToCharArray());
            foreach (string tmpstr in prms)
            {
                string str = tmpstr;
                if (string.IsNullOrEmpty(str) || string.IsNullOrEmpty(str))
                {
                    continue;
                }

                string def = "";
                if (str.Contains("="))
                {
                    def = Convert.ToString(str.Substring(str.IndexOf("=")).Trim());
                }

                if (GetVarType(str) == VarTypes.TypeFloat)
                {
                    str = str.Remove(0, 6);
                    Floats.Add(str, def);
                }
                else if (GetVarType(str) == VarTypes.TypeArray)
                {

                    bool done = false;
                    while (done == false)
                    {
                        if (str.Contains("[") && str.Contains("]"))
                        {
                            str = str.Remove(str.IndexOf("["), (str.IndexOf("]") - str.IndexOf("[")) + 1);
                        }
                        else
                        {
                            done = true;
                        }
                    }

                    Arrays.Add(str, def);
                }
                else if (GetVarType(str) == VarTypes.TypeTagged)
                {
                    _othersVar.Add(str, def);
                }
                else //Its an integer obv..
                {
                    Integers.Add(str, def);
                }
            }
        }

        public List<string> Others(string tag)
        {
            return ((from str in _othersVar.Keys
                let vars = _othersVar[str].Split(new[] {':'}, 2, StringSplitOptions.None)
                select vars).Where(vars => vars[0] == tag).Select(vars => vars[1])).ToList();
        }

        public enum ReturnType
        {
            AsEnum,
            AsString
        }

        public dynamic GetParameterType(string paramName, ReturnType returnType)
        {
            VarTypes ret = GetVarType(paramName);

            if (returnType == ReturnType.AsEnum)
            {
                return ret;
            }

            if (ret == VarTypes.TypeArray)
            {
                return "Array";
            }

            if (ret == VarTypes.TypeFloat)
            {
                return "Float";
            }

            if (ret == VarTypes.TypeInteger)
            {
                return "Integer";
            }

            if (ret == VarTypes.TypeOther)
            {
                return "Other";
            }

            if (ret == VarTypes.TypeTagged)
            {
                return "Tagged";
            }

            return null;
        }
    }
}